### PR TITLE
[Brave News]: Add command line flags for customizing how the V2 feed is generated

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/32175

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_news/browser/feed_v2_knobs.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+#ifndef BRAVE_COMPONENTS_BRAVE_NEWS_BROWSER_FEED_V2_KNOBS_H_
+#define BRAVE_COMPONENTS_BRAVE_NEWS_BROWSER_FEED_V2_KNOBS_H_
+
+namespace brave_news::knobs {
+namespace switches {
+
+// An integer indicating the minimum number of cards (following the hero) in a
+// block.
+extern const char kBraveNewsMinBlockCards[];
```
Comment (fallaciousreasoning) [inline components/brave_news/browser/feed_v2_knobs.h:14]: @LorenzoMinto are there any other knobs you think it'd be a good idea to expose here?

File: components/brave_news/browser/feed_v2_knobs.cc [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+double GetPopRecencyHalfLife() {
+  constexpr double kDefault = 18;
+  return GetDoubleSwitchOrDefault(switches::kBraveNewsPopRecencyHalfLife,
+                                  kDefault);
+}
+
+double GetPopScoreFallback() {
+  constexpr double kDefault = 50;
+  return GetDoubleSwitchOrDefault(switches::kBraveNewsPopRecencyFallback,
```
Comment (sangwoo108) [inline components/brave_news/browser/feed_v2_knobs.cc:85]: Is kBraveNewsPopRecencyFallback missing?

**Review by sangwoo108 — APPROVED** [review]
LGTM 👍

File: components/brave_news/browser/feed_v2_knobs.cc [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+#include "base/strings/string_piece_forward.h"
+
+namespace brave_news::knobs {
+
+namespace {
+
+std::string GetSwitch(const base::StringPiece& switch_name) {
+  return base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII(
+      switch_name);
+}
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_knobs.cc:21]: For now, I think we're okay for them to be command line only - I did have a quick chat to @petemill and he suggested using `base::FeatureParam`, which I didn't know about!

File: components/brave_news/browser/feed_v2_knobs.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+// The boost which is applied to sources where the user is following a channel
+// containing the source. Ideally this is a smaller value than the subscribed
+// boost, as that provides a stronger signal of interest.
+extern const char kBraveNewsChannelSubscribedBoost[];
+
+// The minimum visit weighting to apply to sources (i.e. unvisited sources).
+// This value is used so unvisited sources still show up in the feed. Source
+// visits are calculated as the normalized visit count (i.e. 0 - 1) + this
+// offset.
+extern const char kBraveNewsSourceVisitsMin[];
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_knobs.h:55]: Changed to base::FeatureParam - I think `constexpr` doesn't play nice with component builds (at least on Windows).

File: components/brave_news/browser/feed_v2_knobs.cc (hunk 2 above)
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_knobs.cc:85]: Hahahaha yeah, I'd done a rename which doesn't look like it saved :laughing:

**Comment by fallaciousreasoning** [comment]
@petemill mind double checking if I've done this right?

**Review by sangwoo108 — APPROVED** [review]
Still LGTM 👍

**Pull request merged**